seabug_contracts/CallContract.purs

```diff
@@ -0,0 +1,276 @@
+import Data.BigInt (BigInt)
+import Data.BigInt as BigInt
+import Data.Function (($), (#))
+import Data.List (toUnfoldable)
+import Data.Tuple.Nested ((/\))
+import Data.UInt as UInt
+import Data.Unit (Unit)
+import Effect (Effect)
+import Effect.Aff (error)
+import Effect.Class (liftEffect)
+import Effect.Exception (Error)
+import Metadata.Seabug (SeabugMetadata(SeabugMetadata))
+import Metadata.Seabug.Share (unShare)
+import Seabug.Contract.MarketPlaceBuy (marketplaceBuy)
+import Seabug.Contract.MarketPlaceListNft (ListNftResult, marketPlaceListNft)
+import Seabug.Types (NftCollection(NftCollection), NftData(NftData), NftId(NftId))
+import Serialization.Address (addressBech32, intToNetworkId)
+import Types.Natural as Nat
+import Types.Value (flattenNonAdaValue)
+import UsedTxOuts (newUsedTxOuts)
+import Wallet (mkNamiWalletAff)
+
+-- | Exists temporarily for testing purposes
+callMarketPlaceBuyTest :: String -> Effect (Promise String)
+callMarketPlaceBuyTest str = Promise.fromAff do
+  pure str
+
+-- | Calls Seabugs marketplaceBuy and takes care of converting data types.
+--   Returns a JS promise holding no data.
+callMarketPlaceBuy :: ContractConfiguration -> BuyNftArgs -> Effect (Promise Unit)
+callMarketPlaceBuy cfg args = Promise.fromAff do
+  contractConfig <- buildContractConfig cfg
+  nftData <- liftEffect $ liftEither $ buildNftData args
+  runContract_ contractConfig (marketplaceBuy nftData)
+
+-- | Calls Seabugs marketPlaceListNft and takes care of converting data types.
+--   Returns a JS promise holding nft listings.
+callMarketPlaceListNft :: ContractConfiguration -> Effect (Promise (Array ListNftResultOut))
+callMarketPlaceListNft cfg = Promise.fromAff do
+  contractConfig <- buildContractConfig cfg
+  listnft <- runContract contractConfig marketPlaceListNft
+  pure $ buildNftList <$> listnft
+
+-- | Configuation needed to call contracts from JS.
+type ContractConfiguration =
+  { server_host :: String
+  , server_port :: Int
+  , server_secure_conn :: Boolean
+  , ogmios_host :: String
+  , ogmios_port :: Int
+  , ogmios_secure_conn :: Boolean
+  , datum_cache_host :: String
+  , datum_cache_port :: Int
+  , datum_cache_secure_conn :: Boolean
+  , networkId :: Int
+  }
+
+type BuyNftArgs =
+  { nftCollectionArgs ::
+      { collectionNftCs :: String -- CurrencySymbol
+      , lockLockup :: BigInt -- BigInt
+      , lockLockupEnd :: BigInt -- Slot
+      , lockingScript :: String -- ValidatorHash
+      , author :: String -- PaymentPubKeyHash
+      , authorShare :: BigInt -- Natural
+      , daoScript :: String -- ValidatorHash
+      , daoShare :: BigInt -- Natural
+      }
+  , nftIdArgs ::
+      { collectionNftTn :: String -- TokenName
+      , price :: BigInt -- Natural
+      , owner :: String -- PaymentPubKeyHash
+      }
+  }
+
+-- Placeholder for types I'm not sure how should we represent on frontend.
+type ValueOut = Array { currencySymbol :: String, tokenName :: String, amount :: BigInt }
```

Comment:
I'm not sure if this representation will work. Why not use the browser-builtin `Map` for the non-Ada assets?

Reply:
+ for `Map`.

But note that unlike PS `Data.Map` it can only hold keys of primitive types by value (it holds object keys by reference). So this is not a universal solution.

Reply:
Why do you think this may not work? It basically returns in what is available in Plutus by 
[flattenValue]() :: [Value](https://playground.plutus.iohkdev.io/doc/haddock/plutus-ledger-api/html/Plutus-V1-Ledger-Value.html#t:Value) -> [([CurrencySymbol](https://playground.plutus.iohkdev.io/doc/haddock/plutus-ledger-api/html/Plutus-V1-Ledger-Value.html#t:CurrencySymbol), [TokenName](https://playground.plutus.iohkdev.io/doc/haddock/plutus-ledger-api/html/Plutus-V1-Ledger-Value.html#t:TokenName), [Integer](file:///nix/store/52vknxdwxg37m5wdklm17cfcwj0zxrdd-basement-lib-basement-0.0.12-haddock-doc/share/doc/basement/html/Basement-Compat-Base.html#t:Integer))]

Reply:
I think this will be fine for now, I don't want to hold up moving forward with Seabug